### DISCUSSION TITLE
Keyword handling

### DISCUSSION THREAD

**Discussion by manticore-projects** [prestodb/sql#33] [Ideas]
Coming from JSQLParser, I know that keyword handling in SQL is a nightmare.
As far as I understand it, every Token of the Grammar must be white listened or it becomes a Reserved Keyword automatically.

What I have done for JSQLParser is:
1) Define a set of SQL:2016 based Reserved Keywords
2) Then extract all defined Tokens from the Grammar into a "ALL KEYWORDS" list
3) Then brute-forcefully run tests, where I used "ALL KEYWORDS" as schema, table, column names, aliases, function names etc.
When it fails, I add it to the `Reserved Keywords`. When it succeeds, I add it to the `WhiteList` production in the grammar.

In the result I get a proper documentation of the actual `Reserved Keywords` and the Grammar gets properly adjusted when new tokens or productions are added. 

Caveats:
a) the Tokens are extracted from the Grammar via Regular Expressions, which does not work well for compound tokens.
(I have an implementation using JavaCC, but that would add a dependency.)

b) updating the Grammar semi-automatically works only when the `WhiteList` Production is locked and we agree that it should not be edited manually.

Pros: 
a) less error reports, this has reduced the reported issues by 50 %.
b) new contributors don't need to worry about keyword handling (when they tweak productions and add features)
 
Would something like this be useful for this project? Should I craft a PR? (Its a lot of work so I need to test the waters.)

**Comment by manticore-projects** [comment]
Illustration of my concern:
```sql
SELECT Overlaps( overlaps ) AS overlaps
FROM overlaps.overlaps overlaps
WHERE overlaps = 'overlaps'
    AND (CURRENT_TIME, INTERVAL '1' HOUR) OVERLAPS (CURRENT_TIME, INTERVAL -'1' HOUR)
;
```

From here we go straight to "semantic syntax highlighting" and LSP support :-D

**Reply by kaikalur** [reply]
It's already done elegantly in this grammar. So I don't think it's needed to change that here in this repo.

**Reply by kaikalur** [reply]
For your example, we should give a syntax error as overlaps is not allowed as a keyword (it's a reserved word). So I don't want to allow non-sql compliant syntax here. This is a fully compliant grammar from the spec. For JSQL if you want to do something different, go ahead and do it. But let's keep this as the reference implementation.

**Reply by manticore-projects** [reply]
Thanks for your replay. I can see the `ReservedKeywords` which host around 200 words.
That is far more than the actual SQL:2016 standard defines, I think?

For example there is `SQRT` and I am quite sure that this should not be a reserved keyword?
Same for `OVERLAPS`. It is shown as reserved keyword but allowed for Tablenames, Columns and Functions in any RDBMS I know (see my example above).

Those are not in `NonReservedKeywords`? Am I misunderstanding something wrong?

**Reply by kaikalur** [reply]
Surprisingly, SQRT is a reserved word lol. Similarly OVERLAPS. I cannot post the spec pdf here (due to licensing issues) but if you have access to it, see page 190.

**Reply by kaikalur** [reply]
> in any RDBMS I know (see my example above).

If so, do it in JSQL. We will keep this grammar pure.

**Reply by manticore-projects** [reply]
> If so, do it in JSQL. We will keep this grammar pure.

Thanks for the clarification, I get your point.
So everything in `ReservedKeywords` will be strictly reserved. I needed that for understanding and documentation purpose only.
Cheers.

**Reply by kaikalur** [reply]
> > If so, do it in JSQL. We will keep this grammar pure.
> 
> Thanks for the clarification, I get your point. So everything in `ReservedKeywords` will be strictly reserved. I needed that for understanding and documentation purpose only. Cheers.

Yeah, as you can see this grammar is mechanically generated from the spec :) so I had no idea (until now) OVERLAPS and SQRT (and many others) were reserved keywords.

**Reply by manticore-projects** [reply]
Note to myself:

1) I quickly moved `OVERLAPS` to the `NonReservedKeywords` and this made the Sample above parsing correctly. So the mechanism works in general.
2) We will need to keep a pure Grammar as requested and derive some "practical" Grammars from it (for parsing the actual dirty sql statements roaming the jungle). I need to figure out a smart way to switch between Grammars/Fragments on demand.

**Reply by kaikalur** [reply]
> Note to myself:
> 
> 1. I quickly moved `OVERLAPS` to the `NonReservedKeywords` and this made the Sample above parsing correctly. So the mechanism works in general.

Cool!

> 2. We will need to keep a pure Grammar as requested and derive some "practical" Grammars from it (for parsing the actual dirty sql statements roaming the jungle). I need to figure out a smart way to switch between Grammars/Fragments on demand.

I think you can simply relax the identifier rule to allow reserved words also as identifiers (using a flag) so no need to fork/copy etc. Just a flag that relaxes the spec rules.

**Comment by kaikalur** [comment]
Also, TestKeywords which is part of the standard test suite, does this automatically to make sure when you add a new keyword, it's not reserved word.

**Reply by manticore-projects** [reply]
> Also, TestKeywords which is part of the standard test suite, does this automatically to make sure when you add a new keyword, it's not reserved word.

Thanks for all your support and patience! I do appreciate!

**Reply by manticore-projects** [reply]
Lol, I just broke your test by adding a token to `NonReservedKeywords`. It broke the counter and let the `COUNT` Token test fail! 

Only when I add the Token before the `Count` Token, it works as expected.
Please allow me to modify this particular test and to actually parse the `NonReservedKeywords` -- independent from any order, white space and start/stop tokens.

**Comment by kaikalur** [comment]
That's by design. You should add any no reserved before MAX_NON_RESERVED

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by kaikalur** [comment]
So the recipe here is that reserved and non-reserved words are in two sections and should be contiguous. If there is some bug in the test we should fix it but I don't see the need for the logic to change. We may also need to make the final test (for extraneous keywords) more robust. The test failure is legit.

**Reply by manticore-projects** [reply]
Agreed on the receipt and the sound logic. (I did pretty much the same thing in JSQLParser, just less elegant.)

I would like to suggest the following changes only:
1) instead depending on the start/stop tokens, get the actual tokens from the grammar and feed them into the test
2) test the tokens for more SQL Object identifiers such as schema-, table-, column-names, function names, aliases -- as part of the select items, CTEs, where, order, aggregate functions.
3) write the found/tested tokens onto the Sphinx website 

I will craft a separate PR.

**Comment by kaikalur** [comment]
You can just call the identifier() rule (I think) to test something is accepted as an identifier or not. We don't need separate tests. I like to keep this compact. We want to test only the fact that non-reserved are accepted as identifiers and reserved ones are rejected as identifiers and they are in the right place in the grammar token blocks.

**Comment by kaikalur** [comment]
Actually just add an interface that checks if a given token is a reserved word or not and override it in different parsers to return different values. that's the easiest way
